jupyter/sum-to-zero/sum_to_zero_evaluation.qmd

```diff
@@ -0,0 +1,1558 @@
+extend the BYM2 model to account for disconnected components and singleton nodes.
+
+This has been done by Freni-Sterrantino et al. in 2018 for INLA, and presented in:
+[A note on intrinsic Conditional Autoregressive models for disconnected graphs](https://arxiv.org/abs/1705.04854).
+They provide the following recommendations:
+
+* Non-singleton nodes are given the BYM2 prior
+* Singleton nodes (islands) are given a standard Normal prior
+* Compute per-connected component scaling factor
+* **Impose a sum-to-zero constraint on each connected component**
+
+We have followed these recommendations and implemented this model in Stan.
+The full model is in file `stan/bym2_multicomp.stan`.
+The full implementation details can be found in the notebook
+* [The BYM2_multicomp model in Stan](https://github.com/mitzimorris/geomed_2024/blob/main/python-notebooks/h6_bym2_multicomp.ipynb)
+
+For this case study, we provide 2 implementations of the BYM2_multicomp model:
+one which uses the `sum_to_zero_vector` and one which implements the soft sum-to-zero constraint.
+
+It is necessary to constrain the  the elements of the spatial effects vector `phi` on a component-by-component basis.
+Stan's [slicing with range indexes](https://mc-stan.org/docs/stan-users-guide/multi-indexing.html#slicing-with-range-indexes),
+provides a way to efficiently access each component.
+The helper function `nyc_sort_by_comp_size` both sorts the study data by component and adds the component index to the geodataframe.
+
+In the BYM2 model for a fully connected graph the sum-to-zero constraint on `phi`
+is implemented directly by declaring `phi` to be a `sum_to_zero_vector`, which is a
+[constrained parameter type](https://mc-stan.org/docs/reference-manual/transforms.html#variable-transforms.chapter).
+The declaration:
+```stan
+  sum_to_zero_vector[N] phi;  // spatial effects
+```
+creates a *constrained* variable of length $N$, with a corresponding unconstrained variable of length $N-1$.
+
+In order to constrain slices of the parameter vector `phi`, we do the following:
+
+* In the `parameters` block, we declare the *unconstrained* parameter `phi_raw` as an regular vector `vector` (instead of a `sum_to_zero_vector`).
+    + For a fully connected graph of size $N$, the size of the unconstrained sum-to-zero vector is $N-1$.
+For a disconnected graph with $M$ non-singleton nodes, the size of `phi_raw` is $M$ minus the
+number of connected components.
+
+```stan
+  vector[N_connected - N_components] phi_raw;  // spatial effects
+```
+
+* In the `functions` block, we implement the unconstraining transform.
+
+
+* In the `transformed parameters` block, we apply the constraining transform.
+
+```stan
+  vector[N_connected] phi = zero_sum_components_lp(phi_raw, component_idxs, component_sizes);
+```
+
+The constraining transform is broken into two functions:
+
+* function `zero_sum_constrain_lp`, the actual constraining transform, which corresponds directly
```

Comment:
A couple notes on this:

1. This should probably be a `_jacobian` function (another new 2.36 feature), or, because there is no jacobian adjustment, just a normal function (probably worth mentioning the linearity of the transform here)
2. This function will, [most likely](https://github.com/stan-dev/stanc3/pull/1494), be available built-in in 2.37 under the names `sum_to_zero_constrain` and equivalently `sum_to_zero_jacobian`. Not sure if it's worth mentioning that here or doing a edit later